### PR TITLE
Add new `--proxy-headers` option

### LINKED ISSUE #23431: Allow user to select between `Forwarded` or `X-Forwarded-*` header
### Description

Allow user to explicitly select which header is processed: `Forwarded` or `X-Forwarded-*`. Currently both are processed (in that order).

### Discussion

_No response_

### Motivation

**Background**

Keycloak 21.0.0 introduced support for [RFC7239](https://datatracker.ietf.org/doc/html/rfc7239) `Forwarded` header with following warning ([link](https://www.keycloak.org/docs/latest/release_notes/index.html#support-for-the-standard-forwarded-header-when-running-behind-a-reverse-proxy)):

> Please, make sure your proxy is also overriding the Forwarded header when making requests to Keycloak nodes.

This warning is repeated in [reverse proxy guide](https://www.keycloak.org/server/reverseproxy)

> Take extra precautions to ensure that the client address is properly set by your reverse proxy via the Forwarded or X-Forwarded-For headers. If this header is incorrectly configured, rogue clients can set this header and trick Keycloak into thinking the client is connected from a different IP address than the actual address. This precaution can be more critical if you do any deny or allow listing of IP addresses.

**Problem**

Most proxies, such as NGINX and Envoy only support `X-Forwarded-*`.

Users who previously deployed Keycloak behind proxy and rely on old behavior, will become vulnerable to header forging by upgrading to Keycloak 21.0.0+ if they miss the importance of the note about reconfiguring their proxy to reset the `Forwarded` header.  As a result, external client can set `Forwarded` with any address, and Keycloak will process that over the real client address in `X-Forwarded-*`.

### Details

**Security hardening proposal** 

Instead of depending only on proxy (re)configuration to secure the deployment,  I would like to propose allowing user to explicitly control following Quarkus properties individually: `quarkus.http.proxy.allow-forwarded` vs `quarkus.http.proxy.allow-x-forwarded` ([link](https://quarkus.io/guides/http-reference#reverse-proxy)).

Today the proxy mode enables both as a combination. As a result, one header is automatically picked over another, which may cause fallback to a header set by untrusted client. More predictable behavior would be to avoid automatic fallback and let user to have control over which header is processed. 

Related issue: https://github.com/quarkusio/quarkus/issues/35751

### PR DESCRIPTION
**Pull request by vmuzikar**
Closes #23431

Operator will be handled as a follow-up in https://github.com/keycloak/keycloak/issues/25179.

### CODE REVIEW

File: quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt [hunk 1]
```diff
@@ -150,14 +150,14 @@ HTTP(S):
 --https-protocols <protocols>
                      The list of protocols to explicitly enable. Default: TLSv1.3,TLSv1.2.
 --https-trust-store-file <file>
-                     DEPRECATED: The trust store which holds the certificate information of the
-                       certificates to trust.
+                     DEPRECATED. Use: truststore-paths. -- The trust store which holds the
```
Comment (shawkins) [inline quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt:153]: I'm not sure if it's worth it to convey, but the usage of truststore-paths would be optional as you could just use the defatult location.

File: quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt [hunk 2]
```diff
@@ -150,14 +150,14 @@ HTTP(S):
...
+                       certificate information of the certificates to trust.
 --https-trust-store-password <password>
-                     DEPRECATED: The password of the trust store file.
+                     DEPRECATED. Use: truststore-paths. -- The password of the trust store file.
```
Comment (shawkins) [inline quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt:156]: There will be no replacement for the password - the the store requires a password, the user will have to convert it.

File: quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt [hunk 3]
```diff
@@ -150,14 +150,14 @@ HTTP(S):
...
 --https-trust-store-type <type>
-                     DEPRECATED: The type of the trust store file. If not given, the type is
-                       automatically detected based on the file name. If 'fips-mode' is set to
-                       'strict' and no value is set, it defaults to 'BCFKS'.
+                     DEPRECATED. Use: truststore-paths. -- The type of the trust store file. If not
```
Comment (shawkins) [inline quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt:158]: There is currently no replacement for type - if it's not a pkcs12 nor pem, it's not supported.

**Comment by vmuzikar** [comment]
Adding @ahus1 as a reviewer for the new deprecation logic in docs.

File: quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt [hunk 4]
```diff
@@ -183,8 +183,13 @@ Metrics:
 
 Proxy:
 
---proxy <mode>       The proxy address forwarding mode if the server is behind a reverse proxy.
-                       Possible values are: none, edge, reencrypt, passthrough. Default: none.
+--proxy <mode>       DEPRECATED. Use: proxy-headers. -- The proxy address forwarding mode if the
```
Comment (mabartos) [inline quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt:186]: What about changing the format of the deprecated options to sth like this?
```
DEPRECATED (use 'proxy-headers'): The proxy ...
```

File: services/src/main/java/org/keycloak/truststore/FileTruststoreProviderFactory.java [hunk 5]
```diff
@@ -77,6 +77,10 @@ public void init(Config.Scope config) {
         String policy = config.get(HOSTNAME_VERIFICATION_POLICY);
         String configuredType = config.get("type");
 
+        if (storepath != null || pass != null || configuredType != null) {
+            log.warn("Using deprecated truststore-spi options. Consider using truststore-paths option.");
```
Comment (mabartos) [inline services/src/main/java/org/keycloak/truststore/FileTruststoreProviderFactory.java:81]: I'd be more explicit in this case as we already know the specific provider and format of these options. WDYT?
```suggestion
            log.warn("Using deprecated 'spi-truststore-file-*' options. Consider using 'truststore-paths' option.");
```

I'm ok with not introducing additional complexities here.

File: docs/documentation/upgrading/topics/keycloak/changes-24_0_0.adoc [hunk 6]
```diff
@@ -18,3 +18,23 @@ import AuthZ from 'keycloak-js/authz';
...
 The `tls-hostname-verifier` property should be used instead of the `spi-truststore-file-hostname-verification-policy` property.
+
+= Deprecated `--proxy` option
+
+The `--proxy` option has been deprecated and will be removed in a future release. The following table explains how the deprecated option maps to supported options.
+
+[cols="1,1"]
+|===
+|`kc.sh` (no `proxy` option set)
+|`kc.sh`
```
Comment (mabartos) [inline docs/documentation/upgrading/topics/keycloak/changes-24_0_0.adoc:29]: Is it necessary to have it here? Seems irrelevant to me in the context of `Deprecated --proxy option`. Only users who have specified the option are affected by this deprecation. WDYT?

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMapper.java [hunk 7]
```diff
@@ -96,6 +100,17 @@ ConfigValue getConfigValue(String name, ConfigSourceInterceptorContext context)
...
+        if (config != null && getDeprecatedMetadata().isPresent()) {
+            String logMsg = String.format("Deprecated option in use: \"%s\" from %s.", option.getKey(), config.getConfigSourceName());
+            if (getDeprecatedMetadata().get().getNote() != null) {
+                logMsg += " " + getDeprecatedMetadata().get().getNote();
+            }
+            if (!getDeprecatedMetadata().get().getNewOptionsKeys().isEmpty()) {
+                logMsg += String.format(" Consider using: %s.", String.join(", ", getDeprecatedMetadata().get().getNewOptionsKeys()));
+            }
+            logger.warn(logMsg);
+        }
```
[the remark this answers is not in the document]
Comment (mabartos) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMapper.java:112]: It'd be probably good to store some context about these deprecations and at the end of processing these opts just print it to a console?

**Review by mabartos — COMMENTED** [review]
@vmuzikar I've just briefly gone through the PR and I put some nitpicks comments below. I'll make more deep dive review later. 

It'd be good to separate the work into 2 commits related to the `proxy-headers` and deprecation logic for better review and manageability. WDYT?

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMapper.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (pedroigor) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMapper.java:112]: Similarly as we do for errors, right? I think it will bring more consistency, yeah ...

Perhaps as clear entry like:

```
You are using deprecated options:

* option-foo: <note>
* option-bar: <note>

These options will be removed in the feature and no longer available from future release.
```

Something better than this ...

[the remark this answers is not in the document]
Comment (pedroigor) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMapper.java:112]: If we want to be really strict about it, we could even fail the build/start commands and introduce a `--use-deprecated` option to make it really painful and force people to migrate. As well as force us to take real care when designing options and deprecating them :)

[the remark this answers is not in the document]
Comment (vmuzikar) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMapper.java:112]: Aggregating the log message to include all used deprecated options sounds good. Next week I'll try to figure out the way to do it.

File: quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt (hunk 2 above)
Comment (vmuzikar) [reply quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt:156]: I'll add a more descriptive note.

File: quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt (hunk 1 above)
Comment (vmuzikar) [reply quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt:153]: I'll add a more descriptive note there (`deprecatedWithNote`).

File: quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt (hunk 3 above)
Comment (vmuzikar) [reply quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt:158]: I'll add a more descriptive note.

File: docs/documentation/upgrading/topics/keycloak/changes-24_0_0.adoc (hunk 6 above)
Comment (vmuzikar) [reply docs/documentation/upgrading/topics/keycloak/changes-24_0_0.adoc:29]: I wanted to make the comparison as explicit as possible, incl. how no proxy maps to no proxy headers. Does it make any sense? :)

**Comment by vmuzikar** [comment]
Thanks for the reviews!

> It'd be good to separate the work into 2 commits related to the `proxy-headers` and deprecation logic for better review and manageability. WDYT?

@mabartos Since the `proxy` option was the first major candidate for deprecation (the truststore PR was merged just yesterday), it felt natural to keep it together to actually see the deprecation logic in action. I can split it now (one PR for the deprecation logic applied to the mTLS options, and second for the proxy options) if needed. But IMHO the size of this PR is manageable so far.

**Review by ahus1 — CHANGES_REQUESTED** [review]
@vmuzikar - thank you for these changes, they make sense to me and I understand the docs. 

I pushed one change to the table, as IMHO a table should have a heading. If you use AsciiDoc formatting in table cells, the cells should be marked with `a` which I added. 

With the new option, there are still two places where the now deprecated syntax is used in the docs. 

See:

* `reverseproxy.adoc`: section "Trust the proxy to set hostname"
* `hostname.adoc`: two usages, search for `--proxy`.

I assume all three occurrences should use the new syntax.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMapper.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (vmuzikar) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMapper.java]: Reworked that so the deprecation log message is not printed only once and aggregates all deprecated options in use.

I would not be too strict (require the users to explicitly set `--use-deprecated true`) as deprecating something should not itself be a breaking change requiring you to set a new switch.

File: quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt (hunk 1 above)
Comment (vmuzikar) [reply quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt]: Change the wording a bit and removed the hinted option.

File: quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt (hunk 2 above)
Comment (vmuzikar) [reply quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt]: Change the wording a bit and removed the hinted option.

File: quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt (hunk 3 above)
Comment (vmuzikar) [reply quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt]: Change the wording a bit and removed the hinted option.

**Comment by vmuzikar** [comment]
> I pushed one change to the table, as IMHO a table should have a heading. If you use AsciiDoc formatting in table cells, the cells should be marked with `a` which I added.

@ahus1 Thank you! Added a comment, not sure if the changes work as expected.

> With the new option, there are still two places where the now deprecated syntax is used in the docs.

Very good catch! Thank you.

**Comment by vmuzikar** [comment]
@shawkins @pedroigor @mabartos @ahus1 I addressed the review comments. Feel free to give it another look when you get a chance.

**Comment by vmuzikar** [comment]
DCO is failing due to missing signature on the accepted suggestion commit. We can override the check on merge, or I can squash the commits manually before merging.

File: docs/guides/server/reverseproxy.adoc [hunk 8]
```diff
@@ -8,11 +8,31 @@ title="Using a reverse proxy"
...
+For example:
+
+<@kc.start parameters="--proxy-headers forwarded"/>
+
+WARNING: If either `forwarded` or `xforwarded` is selected, make sure your reverse proxy properly sets and overwrites the `Forwarded` or `X-Forwarded-*` headers respectively. To set these headers, consult the documentation for your reverse proxy. Misconfiguration will leave {project_name} exposed to security vulnerabilities.
+
+Take extra precautions to ensure that the client address is properly set by your reverse proxy via the `Forwarded` or `X-Forwarded-For` headers.
+If this header is incorrectly configured, rogue clients can set this header and trick {project_name} into thinking the client is connected from a different IP address than the actual address. This precaution can be more critical if you do any deny or allow listing of IP addresses.
+
+Some {project_name} features rely on the assumption that the remote address of the HTTP request connecting to {project_name} is the real IP address of the clients machine.
```
[the remark this answers is not in the document]
Comment (vmuzikar) [reply docs/guides/server/reverseproxy.adoc]: Removed the sentence.

File: docs/documentation/upgrading/topics/keycloak/changes-24_0_0.adoc [hunk 9]
```diff
@@ -18,3 +18,26 @@ import AuthZ from 'keycloak-js/authz';
...
 
 The `tls-hostname-verifier` property should be used instead of the `spi-truststore-file-hostname-verification-policy` property.
+
+= Deprecated `--proxy` option
+
+The `--proxy` option has been deprecated and will be removed in a future release. The following table explains how the deprecated option maps to supported options.
+
+[%autowidth,cols="a,a"]
+|===
+| Deprecated usage | New usage
```
[the remark this answers is not in the document]
Comment (ahus1) [reply docs/documentation/upgrading/topics/keycloak/changes-24_0_0.adoc:28]: I ran the following command: 

```
mvn clean install -am -pl docs/documentation/upgrading -Pdocumentation
``` 

This outputs a file-link on the console. Once I click on it, I see the following which is with the headline: 

image

[the remark this answers is not in the document]
Comment (vmuzikar) [reply docs/documentation/upgrading/topics/keycloak/changes-24_0_0.adoc:28]: Thanks! Today, I was forgetting the `-Pdocumentation`. (Last time when I wrote the docs I included it.)

File: docs/guides/server/hostname.adoc [hunk 10]
```diff
@@ -133,7 +133,7 @@ In this example, the server is accessible using a port other than the default po
 In this example, the server is running behind a proxy and both the server and the proxy are using their own certificates, so the communication between {project_name} and the proxy is encrypted. Because we want the proxy to use its own certificate, the proxy mode `reencrypt` will be used. We need to keep in mind that the proxy configuration options (as well as hostname configuration options) are not changing the ports on which the server actually is listening on (it changes only the ports of static resources like JavaScript and CSS links, OIDC well-known endpoints, redirect URIs, etc.). Therefore, we need to use HTTP configuration options to change the {project_name} server to internally listen on a different port, e.g. 8543. The proxy will be listening on the port 8443 (the port visible while accessing the console via a browser). The example hostname `my-keycloak.org` will be used for the server and similarly the admin console will be accessible via the `admin.my-keycloak.org` subdomain.
```
Comment (ahus1) [inline docs/guides/server/hostname.adoc:133]: This is also referring to proxy mode `reencrypt`, and should be updated.

File: docs/guides/server/hostname.adoc [hunk 11]
```diff
@@ -133,7 +133,7 @@ In this example, the server is accessible using a port other than the default po
...
 
 .{project_name} configuration:
-<@kc.start parameters="--proxy=reencrypt --https-port=8543 --hostname-url=https://my-keycloak.org:8443 --hostname-admin-url=https://admin.my-keycloak.org:8443"/>
+<@kc.start parameters="--proxy-headers=forwarded|xforwarded --https-port=8543 --hostname-url=https://my-keycloak.org:8443 --hostname-admin-url=https://admin.my-keycloak.org:8443"/>
 
 Note: there is currently no difference between the `passthrough` and `reencrypt` modes. For now, this is meant for future-proof configuration compatibility. The only difference is that when the `edge` proxy mode is used, HTTP is implicitly enabled (again as mentioned above, this does not affect the server behaviour).
```
Comment (ahus1) [inline docs/guides/server/hostname.adoc:138]: This paragraph needs to be changed to match the updated CLI above, as it is still talking about `passthrough` and `reencrypt`.

**Review by ahus1 — CHANGES_REQUESTED** [review]
Sorry, found two paragraph before/after the CLI which should also change - probably they can be a lot shorter now?

**Review by ahus1 — DISMISSED** [review]
Thank you, looks good!

File: docs/guides/templates/options.adoc [hunk 12]
```diff
@@ -22,6 +22,12 @@
 *CLI:* `${option.keyCli}` +
 *Env:* `${option.keyEnv}`
 --
+
+<#if option.deprecated?has_content>
+*DEPRECATED.*
+${option.deprecated.note!}<#if option.deprecated.newOptionsKeys?has_content><#if option.deprecated.note?has_content> </#if>Use: <#list option.deprecated.newOptionsKeys as key>`+${key}+`<#if key?has_next>, </#if></#list>.</#if>
+</#if>
+
```
Comment (mabartos) [inline docs/guides/templates/options.adoc:30]: Works as expected.

image

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/cli/Picocli.java [hunk 13]
```diff
@@ -307,21 +313,48 @@ public static void validateNonCliConfig(List<String> cliArgs, AbstractCommand ab
...
         if (!ignoredBuildTime.isEmpty()) {
-            outputIgnoredProperties(ignoredBuildTime, true, out);
+            outputIgnoredProperties(ignoredBuildTime, true, logger);
         } else if (!ignoredRunTime.isEmpty()) {
-            outputIgnoredProperties(ignoredRunTime, false, out);
+            outputIgnoredProperties(ignoredRunTime, false, logger);
+        }
+
+        if (!deprecatedInUse.isEmpty()) {
+            logger.warn("The following used options are DEPRECATED and will be removed in a future release:\n" + String.join("\n", deprecatedInUse));
```
Comment (mabartos) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/cli/Picocli.java:350]: Works as expected (provided `--cache` and `--https-trust-store-password`).

Small improvement would be to get rid of the KC namespace as we don't deprecate any other props. For property config file source the current state is relevant, but for CLI not so much. But, it's only a nitpicking.   
```
2023-12-11 12:10:03,314 WARN  [org.keycloak.quarkus.runtime.cli.Picocli] (main) The following used options are DEPRECATED and will be removed in a future release:
        - kc.https-trust-store-password: Use the System Truststore instead, see the docs for details.
        - kc.proxy: Consider using proxy-headers.
```

File: quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt [hunk 14]
```diff
@@ -150,14 +150,17 @@ HTTP(S):
 --https-protocols <protocols>
                      The list of protocols to explicitly enable. Default: TLSv1.3,TLSv1.2.
 --https-trust-store-file <file>
-                     DEPRECATED: The trust store which holds the certificate information of the
-                       certificates to trust.
+                     DEPRECATED. Use the System Truststore instead, see the docs for details. --
```
Comment (pedroigor) [inline quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt:153]: Is it expected the `--` here?

File: quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt [hunk 15]
```diff
@@ -150,14 +150,17 @@ HTTP(S):
...
+                       The trust store which holds the certificate information of the certificates
+                       to trust.
 --https-trust-store-password <password>
-                     DEPRECATED: The password of the trust store file.
```
Comment (pedroigor) [inline quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt:156]: I think the option description is a bit confusing. Perhaps:

```
DEPRECATED: The password of the trust store file.  Use the System Truststore instead, see the docs for details..
```

**Review by mabartos — DISMISSED** [review]
@vmuzikar Nice work! LGTM.

File: quarkus/config-api/src/main/java/org/keycloak/config/OptionBuilder.java [hunk 16]
```diff
@@ -111,12 +114,32 @@ public List<String> get() {
         return this;
     }
 
+    public OptionBuilder<T> deprecated() {
+        this.deprecatedMetadata = new DeprecatedMetadata();
+        return this;
+    }
+
+    public OptionBuilder<T> deprecatedWithNote(String note) {
```
Comment (pedroigor) [inline quarkus/config-api/src/main/java/org/keycloak/config/OptionBuilder.java:122]: Isn't enough the `deprecatedWithNote(String, String...)` method?

Also can we just name it `deprecated`? I mean, `WithNote` seems redundant as the parameter name is kinda making clear about setting a `note`.

To make easier to set deprecation without a note but the new options we just use a `deprecate(Set<Option>)`. So you have:

```
builder.deprecate(Set.of(OptionA, OptionB));
```

File: docs/guides/server/hostname.adoc [hunk 17]
```diff
@@ -40,7 +40,7 @@ Most of the time, it should be enough to set the `hostname` option in order to c
 When using the `hostname` option the server is going to resolve the HTTP scheme, port, and path, automatically so that:
 
 * `https` scheme is used unless you set `hostname-strict-https=false`
-* if the `proxy` option is set, the proxy will use the default ports (i.e.: 80 and 443).  If the proxy uses a different port, it needs to be specified via the `hostname-port` configuration option
+* if the `proxy-headers` option is set, the proxy will use the default ports (i.e.: 80 and 443).  If the proxy uses a different port, it needs to be specified via the `hostname-port` configuration option
```
Comment (pedroigor) [inline docs/guides/server/hostname.adoc:43]: IMO, we should not recommend using `hostname-port` but using the `hostname-url` option.

This option does not much sense if you can just provide the URL you want to expose the frontend URLs (same thing for admin). The `hostname-port` is kinda another option causing the confusion you want to solve with the on-going changes in the hostname configuration and implementation.

File: quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt (hunk 14 above)
Comment (vmuzikar) [reply quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt]: Rewrote the message a bit. Now it looks like
Screenshot 2023-12-11 at 15 08 00
Screenshot 2023-12-11 at 14 37 03

File: quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt (hunk 15 above)
Comment (vmuzikar) [reply quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt:156]: Rewrote the message a bit.

File: quarkus/config-api/src/main/java/org/keycloak/config/OptionBuilder.java (hunk 16 above)
Comment (vmuzikar) [reply quarkus/config-api/src/main/java/org/keycloak/config/OptionBuilder.java]: Done.

File: docs/guides/server/hostname.adoc (hunk 17 above)
Comment (vmuzikar) [reply docs/guides/server/hostname.adoc]: Done.

**Review by ghost — COMMENTED** [review]
Unreported flaky test detected, please review

**Review by mabartos — DISMISSED** [review]
@vmuzikar LGTM

**Comment by mabartos** [comment]
@vmuzikar Oops, conflict.

**Review by pedroigor — DISMISSED** [review]

**Comment by vmuzikar** [comment]
> Oops, conflict.

Rebased, thanks.

**Review by pedroigor — DISMISSED** [review]

**Review by mabartos — DISMISSED** [review]

**Comment by mabartos** [comment]
@vmuzikar @pedroigor Resolved conflicts and rebased.

**Review by mabartos — APPROVED** [review]

**Review by shawkins — APPROVED** [review]

**Review by abstractj — APPROVED** [review]
